src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -342,68 +359,57 @@ void symbol_factoryt::gen_nondet_size_array_init(
   auto const &size_symbol = new_tmp_symbol(size_type(), "size");
 
   auto const &element_type = array.type().subtype();
-  code_ifthenelset initialization_if_then_else;
-  auto *current_if_then_else = &initialization_if_then_else;
+  const bool is_string = is_string_array(array);
 
-  for(std::size_t i = 1; i < max_array_size; ++i)
-  {
-    auto const &array_size = from_integer(i, size_type());
-    auto array_type = array_typet{element_type, array_size};
-
-    auto &static_array = new_tmp_symbol(
-      array_type, id2string(array_name) + "_" + std::to_string(i));
-    static_array.is_static_lifetime = true;
-
-    const constant_exprt &size_expr = array_size;
-
-    current_if_then_else->cond() = binary_exprt(
-      size_cond_symbol.symbol_expr(), ID_lt, size_expr, bool_typet{});
-    code_blockt then_case;
-    then_case.add(code_assignt(size_symbol.symbol_expr(), size_expr));
-    gen_nondet_array_init(
-      then_case, static_array.symbol_expr(), depth, recursion_set);
-    then_case.add(code_assignt(
-      array,
-      address_of_exprt(
-        index_exprt(
-          static_array.symbol_expr(),
-          from_integer(0, size_type()),
-          array_type.subtype()),
-        pointer_type(array_type.subtype()))));
-    current_if_then_else->then_case() = then_case;
-    if(i + 1 < max_array_size)
-    {
-      current_if_then_else->else_case() = code_ifthenelset{};
-      current_if_then_else = reinterpret_cast<code_ifthenelset *>(
-        &current_if_then_else->else_case());
-    }
-    else
-    {
+  auto initialize_nth_array =
+    [&](std::size_t current_array_size, code_blockt &block) {
```

Comment:
This makes for a fairly large lambda... Should it be a helper function instead? At the very least it could do with a few comments to document...

Reply:
I'll have another look at that, see if I can make this nicer.